### ISSUE TITLE
Exception in thread Thread-1:

### ISSUE THREAD

**Issue by KristianDukov** [danpaquin/gdax-python#249]
Hello all,

I am trying to extract only the best_ask from the messages but it's not working as expected. Here is my code

`from __future__ import print_function
import json

from threading import Thread
from websocket import create_connection, WebSocketConnectionClosedException

class WebsocketClient(object):
    def __init__(self, url="wss://ws-feed.gdax.com", products=None, message_type="subscribe"):
        self.url = url
        self.products = products
        self.type = message_type
        self.stop = False
        self.ws = None
        self.thread = None

    def start(self):
        def _go():
            self._connect()
            self._listen()

        self.on_open()
        self.thread = Thread(target=_go)
        self.thread.start()

    def _connect(self):
        if self.products is None:
            self.products = ["BTC-USD"]
        elif not isinstance(self.products, list):
            self.products = [self.products]

        if self.url[-1] == "/":
            self.url = self.url[:-1]

        self.ws = create_connection(self.url)

        self.stop = False
        sub_params = {"type":"subscribe","session_token":"5834befb43df3ea6ebf0da574ca7710209b718237a5910263afb7fff726ebc13","channels":[{"name":"ticker_1000","product_ids":["BTC-EUR","ETH-BTC","ETH-EUR"]}]}
        self.ws.send(json.dumps(sub_params))
        if self.type == "heartbeat":
            sub_params = {"type": "heartbeat", "on": True}
            self.ws.send(json.dumps(sub_params))

    def _listen(self):
        while not self.stop:
            try:
                msg = json.loads(self.ws.recv())
            except Exception as e:
                self.on_error(e)
            else:
                self.on_message(msg)

    def close(self):
        if not self.stop:
            if self.type == "heartbeat":
                self.ws.send(json.dumps({"type": "heartbeat", "on": False}))
            self.on_close()
            self.stop = True
            try:
                if self.ws:
                    self.ws.close()
            except WebSocketConnectionClosedException as e:
                pass

    def on_open(self):
        print("-- Subscribed! --\n")

    def on_close(self):
        print("\n-- Socket Closed --")

    def on_message(self, msg):
        print(msg['type'])

    def on_error(self, e):
        return

if __name__ == "__main__":
    import gdax
    import time

    class MyWebsocketClient(gdax.WebsocketClient):
        def on_open(self):
            self.url = "wss://ws-feed.gdax.com/"
            self.message_count = 0
            print("Let's count the messages!")

        def on_message(self, msg):
            print(msg['best_ask'])
            print(msg['best_bid'])
            self.message_count += 1

    wsClient = MyWebsocketClient()
    wsClient.start()
    # Do some logic with the data
    while wsClient.message_count < 10:
        print("\nMessageCount =", "%i \n" % wsClient.message_count)
        time.sleep(1)

    wsClient.close()`

And here is the output:

`Let's count the messages!

MessageCount = 0 

6858.5
0.1004
690
Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Users\owner\AppData\Local\Programs\Python\Python36-32\lib\threading.py", line 916, in _bootstrap_inner
    self.run()
  File "C:\Users\owner\AppData\Local\Programs\Python\Python36-32\lib\threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\owner\AppData\Local\Programs\Python\Python36-32\lib\site-packages\gdax\websocket_client.py", line 24, in _go
    self._listen()
  File "C:\Users\owner\AppData\Local\Programs\Python\Python36-32\lib\site-packages\gdax\websocket_client.py", line 55, in _listen
    self.on_message(msg)
  File "websocket_client.py", line 92, in on_message
    print(msg['best_ask'])
KeyError: 'best_ask'

MessageCount = 3 `

Why it works only once and then throws an error? How can I fix it?

Thank you

**Comment by alimcmaster1** [comment]
Have you looked to see if the message you receive from the socket contains that key?

**Comment by KristianDukov** [comment]
Yes. Here is an example of a message

`{'type': 'ticker', 'sequence': 3356515171, 'product_id': 'BTC-EUR', 'price': '6779.33000000', 'open_24h': '6740.00000000', 'volume_24h': '3208.08039459', 'low_24h': '6430.00000000', 'high_24h': '7052.54000000', 'volume_30d': '166993.05547426', 'best_bid': '6781.89', 'best_ask': '6781.9'}
{'type': 'ticker', 'sequence': 898842414, 'product_id': 'ETH-BTC', 'price': '0.10059000', 'open_24h': '0.10000000', 'volume_24h': '4591.65270795', 'low_24h': '0.09943000', 'high_24h': '0.10085000', 'volume_30d': '603591.47429186', 'best_bid': '0.10058', 'best_ask': '0.10059'}
{'type': 'ticker', 'sequence': 577571644, 'product_id': 'ETH-EUR', 'price': '683.83000000', 'open_24h': '675.20000000', 'volume_24h': '18447.74589385', 'low_24h': '635.66000000', 'high_24h': '708.50000000', 'volume_30d': '1116479.47465227', 'best_bid': '683.83', 'best_ask': '683.84'}
{'type': 'subscriptions', 'channels': [{'name': 'ticker_1000', 'product_ids': ['BTC-EUR', 'ETH-BTC', 'ETH-EUR']}]}`

**Comment by danpaquin** [comment]
Hello Kristian,

Thank you for your question.  This is not an issue with the wrapper, so I will close it out for now.  If this error continues to be unique to this project, please open a new issue.  I have proposed two solutions for you below.

You could try to check (1) if msg['type'] == 'ticker' or (2) if 'best_bid' in msg.  This should resolve the issue of GDAX possibly sending incomplete packets, etc.
